tools/cwrap/plugins/WrapDim.py

```diff
@@ -0,0 +1,45 @@
+from . import CWrapPlugin
+from string import Template
+
+
+class WrapDim(CWrapPlugin):
+
+    DIM_WRAP_NDIM_TEMPLATE = Template(
+        """${arg_tensor}->nDimension""")
+
+    DIM_WRAP_CHECK_TEMPLATE = Template(
+        """THPUtils_assert(${arg_to_wrap} >= -(${ndim}) && ${arg_to_wrap} < (${ndim}),
+         "dimension out of range (expected to be in range of [%d, %d], but got %d)",
+         -(${ndim}), (${ndim})-1, ${arg_to_wrap})""")
+
+    DIM_WRAP_CODE_TEMPLATE = Template(
+        """if (${arg_to_wrap} < 0) ${arg_to_wrap} += (${ndim});""")
+
+    def initialize(self, cwrap):
+        self.cwrap = cwrap
+
+    def process_option_code_template(self, template, option):
+        new_code = []
+        for i, arg in enumerate(option['arguments']):
+            if 'wrap_dim' not in arg:
+                continue
+
+            params = arg.get('wrap_dim').split("+")
+            arg_tensor = params[0]
+
+            arg_tensor = "arg_" + arg_tensor
+            arg_to_wrap = "arg_" + arg.get('assign_name', arg['name'])
```

Comment:
Right now yes because we only use `assign_name` for stateless implementation (excluding genericNN that works completely differently).
But if some other plugin start to use `assign_name`, this code would break, so I think its better to keep it like that.